Move Meter Growth Into Effects

geoffthemedio opened the pull request:
Move current meter growth from C++ into scripted effects.

geoffthemedio committed:
Move PopCenter current meter changes into effects from C++.

geoffthemedio committed:
-Don't calculate discrepancies for paired active meters.
-grooming

geoffthemedio committed:
Removed DualMeter::CurrentProjectedTarget and just got the values directly.

geoffthemedio committed:
-Moved ResourceCenter meters into effects.
-Moved current / active meter effects later in meter priorities.
-Currently using a few different formats for current meter effects, to test for differences.

geoffthemedio committed:
grooming

geoffthemedio committed:
Replaced Planet::m_just_conquered with Planet::m_turn_last_conquered.

geoffthemedio committed:
Replaced boolean tracking if a planet has been recently invaded with a number tracking when it was last invaded, along with FOCS API exposure.

geoffthemedio committed:
grooming

geoffthemedio committed:
Exposed LastTurnAttackedByShip and LastTurnConquered to Python API.

geoffthemedio committed:
grooming

geoffthemedio committed:
Added ImmediateValueVariable ValueRef variation that, for meters, returns the current value instead of the initial value, which is returned by the standard ValueRef::Variable::Eval. Only useful and implemented for meter values.

geoffthemedio committed:
Made objects list column show current rather than initial meter values.

geoffthemedio committed:
Removed SidePanel::PlanetPanel::FocusDropListOpened, the execution of which was causing the turn a planet's focus was changed to lose track of the original focus and the turn it was last changed, preventing undoing a focus change from reverting the tracking of the turn of the last change.

geoffthemedio committed:
Removed ImmediateValueVar and replace with (work in progress) parameter to ValueRef::Variable.

geoffthemedio committed:
Debug logging.

geoffthemedio committed:
trivial grooming

geoffthemedio committed:
Added missing meter parser.

geoffthemedio committed:
-Added Value() wrapper around object meter ValueRefs, which will return the immediate/current value of the meter, rather than the turn-initial value that is returned otherwise.
-grooming

geoffthemedio committed:
Added LastTurnAttackedByShip ValueRef.

geoffthemedio committed:
-Moved additional meter adjustments from C++ into FOCS.
-Wrapped many meter references with Value() when setting active meter values. This means that target or max meter value references in active meter setting effects will use immediate target or max meter values, rather than initial values.

geoffthemedio committed:
Optimizing binary ValueRef::Operation evaluation.

geoffthemedio committed:
Reworked construction growth/decay effect.

agrrr3 reviewed (changes requested):
Cant comment directly on the relevant source snippets it seems....

Decreasing stockpile limit is missing
STANDARD_STOCKPILE_GROWTH

I also find the formula for in/decrease confusing and not consistently used (e.g. different in construction).
Also dont see how force-energy structures could work with the current code.
Maybe a meters file which provides a macro for in/decrease for consistency and easier changeability?

geoffthemedio committed:
Removed duplicate logging line.

geoffthemedio committed:
Reimplemented force-energy structure resource and construction meter changes with immediate meter Value references.

geoffthemedio committed:
-Added MaxStockpile and Stockpile bound valueref parsers.
-Removed TradeStockpile ValueRef.

geoffthemedio committed:
Added stockpile meter growth macros.

geoffthemedio committed:
trivial grooming

geoffthemedio committed:
Changed DoubleToString to round to nearest instead of round down.

geoffthemedio committed:
-Addeded parameter to ColouredNumber to control number of digits.
-grooming

geoffthemedio committed:
grooming

dbenage-cx reviewed (commented):
Seem to be getting slightly more output than target output when max is steady. e.g. Target Industry 10.0 but Industry Output is 10.2.

Planet shields/defense do not drop below 1.0 when attacked, and do not increase from 0.0 on outposts.

Troops continue to increase on a blockaded/bombarded planet.

geoffthemedio committed:
Reverted misleading movement of accounting labels to newlines.

geoffthemedio committed:
Removed line missed during rebase conflict resolution.

geoffthemedio committed:
Added once-per-turn default shield regen to root defense tech.

geoffthemedio committed:
Moved shield growth effects into separate effectsgroups with conditions requiring the planet to have not been attacked in the previous turn.

geoffthemedio committed:
trivial grooming

geoffthemedio committed:
Tweak format of value comparison condition.

geoffthemedio committed:
Fixed / updated shield and defense setting effects.

geoffthemedio committed:
trivial grooming

geoffthemedio committed:
Added extra meter update and backprojection when creating universe to resolve some first-turn "unknown" contributions to meter values.

geoffthemedio committed:
grooming

geoffthemedio committed:
Added braces to disambiguate if-else relation.

geoffthemedio committed:
-Made planet shield growth effect only apply to unowned planets, to avoid duplicating a tech effect.
-Added some not-attacked-last-turn conditions to planet meter growth.

geoffthemedio committed:
grooming, mostly auto replacement

geoffthemedio committed:
Fixed poorly edited copy-paste.

geoffthemedio committed:
Effects to increase meter values up to maxes.

geoffthemedio committed:
minor grooming

geoffthemedio committed:
Wrap immediate value ValueRef::Variable::Dump output in "Value( ... )" to match parser format.

geoffthemedio committed:
Fixed use of Target reference in scope condition (at which point the target object is not yet defined).

geoffthemedio committed:
-Separated target/max and active meter effects with priorities.
-Moved lifecycle manipulating effects into macro.

geoffthemedio committed:
Added a few system names.

geoffthemedio committed:
Added trace effects logging of objects received by human client at start of turn.

geoffthemedio committed:
grooming

geoffthemedio committed:
-Added function to reset objects' meters.
-Removed unused / broken ApplyMeterEffectsAndUpdateTargetMaxUnpairedMeters variant.
-Based meter discrepancies on difference between first receive and recalculated current values, instead of difference between current and initial.
-Added UpdateMeterEstimates variants that can have their effect accounting turned on/off.
-grooming

geoffthemedio committed:
-Added extra meter update in server before sending turn updates, to make client and server target meter estimates consistent.
-Moved some logging into trace effects filter.

geoffthemedio committed:
Removed scope condition that may have caused inconsistent population setting effect execution, replaced with formula modification.

geoffthemedio committed:
Moved outpost supply increase effect into SHP_GAL_EXPLO with other supply-setting effects.

geoffthemedio committed:
Uncommeted and changed debug line to trace line.

geoffthemedio committed:
Fixed Barrier Shield scripting errors.

geoffthemedio committed:
Added some macro includes, as apparently the *-based includes at higher level aren't working.

geoffthemedio committed:
Replaced ThisSpecies use in colony building scripts with macro parameter.

geoffthemedio committed:
Switched planet troop strength during ground combat to use initial instead of current value, so that values shown to players during the preceeding turn should be accurate.

geoffthemedio committed:
Made troop display on invade button use initial rather than current meter value to be consistent with what is used in ground combat resolution on server.

geoffthemedio committed:
Removed seemingly out of date / misleading comment.

geoffthemedio committed:
Updated copy-paste error.

geoffthemedio committed:
Make minimum digits with DoubleToString 3, to avoid quirky issues with 2 or fewer digits.

dbenage-cx reviewed (commented):
Otherwise LGTM

geoffthemedio committed:
-Removed unnecessary (and badly copy-pasted) condition when growing / reducing happiness.
-Added endlines to ends of script files.
